src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.h

```diff
@@ -0,0 +1,199 @@
+        {
+            angle_to_block = angle_to_positive_goalpost + shot_angle_increment * 5.0;
+        }
+
+        // Shot ray to block
+        Ray ray(enemy_threat_origin, angle_to_block);
+
+        return findDefenseAreaIntersection(field, ray, robot_obstacle_inflation_factor);
+    }
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto block_threat_s = state<BlockThreatState>;
+
+        // update_e is the _event_ that the CreaseDefenderFSM responds to
+        const auto update_e = event<Update>;
+
+        /**
+         * This is an Action that blocks the threat
+         *
+         * @param event CreaseDefenderFSM::Update event
+         */
+        const auto block_threat = [this](auto event) {
+            Point destination       = event.common.robot.position();
+            auto block_threat_point = findBlockThreatPoint(
+                event.common.world.field(), event.control_params.enemy_threat_origin,
+                event.control_params.crease_defender_alignment,
+                event.control_params.robot_obstacle_inflation_factor);
+            if (block_threat_point)
+            {
+                destination = block_threat_point.value();
+            }
+            else
+            {
+                LOG(WARNING)
+                    << "Could not find a point on the defense area to block a potential shot";
+            }
+            Angle face_threat_orientation =
+                (event.control_params.enemy_threat_origin - event.common.robot.position())
+                    .orientation();
+
+            // Chip to the enemy half of the field
+            double chip_distance = event.common.world.field().xLength() / 3.0;
+            // If enemy threat is on the sides, then chip to near the edge of the field
+            if (event.control_params.enemy_threat_origin.x() <
+                event.common.world.field().friendlyDefenseArea().xMax())
+            {
+                chip_distance = event.common.world.field().yLength() / 3.0 -
+                                event.common.world.field().friendlyDefenseArea().yMax();
+            }
+
+            event.common.set_intent(std::make_unique<MoveIntent>(
+                event.common.robot.id(), destination, face_threat_orientation, 0.0,
+                DribblerMode::OFF, BallCollisionType::ALLOW,
+                AutoChipOrKick{AutoChipOrKickMode::AUTOCHIP, chip_distance},
+                MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0));
+        };
+
+        /**
+         * This guard is used to check if the robot is done moving
+         *
+         * @param event CreaseDefenderFSM::Update event
+         *
+         * @return if robot has reached the destination
+         */
+        const auto ball_in_enemy_half = [](auto event) {
+            return contains(event.common.world.field().enemyHalf(),
+                            event.common.world.ball().position());
```

Comment:
I'm a bit confused how the ball being in the enemy half is related to the robot reaching its destination

Reply:
mb, I changed it to use the MoveFSM so it's now about reaching destination